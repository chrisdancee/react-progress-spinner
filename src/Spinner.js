import styles from "./Spinner.module.css";

const Spinner = ({ percentage = 0, radius = 200, thickness = 12 }) => {
  const circumference = 2 * Math.PI * radius;
  const offset = circumference - circumference / (100 / percentage);
  const padding = thickness / 2;

  return (
    <div
      className={`${styles.container} react-progress-spinner-container`}
      style={{
        fontSize: `${radius}px`,
      }}
      role="progressbar"
      aria-busy={percentage < 100}
      aria-live="assertive"
      aria-valuenow={percentage}
      data-testid="spinner"
    >
      <svg
        className={`${styles.spinner} react-progress-spinner-spinner`}
        height={(radius + padding) * 2}
        viewBox={`0 0 ${(radius + padding) * 2} ${(radius + padding) * 2}`}
        width={(radius + padding) * 2}
        aria-hidden="true"
      >
        <circle
          className={`${styles.track} react-progress-spinner-spinner`}
          r={radius + padding}
          cx={radius + padding}
          cy={radius + padding}
        />
        <circle
          className={`${styles.progress} react-progress-spinner-progress`}
          r={radius}
          cx={radius + padding}
          cy={radius + padding}
          style={{
            strokeDasharray: circumference,
            strokeDashoffset: offset,
            strokeWidth: thickness,
          }}
        />
        <circle
          className={`${styles.inset} react-progress-spinner-inset`}
          r={radius - padding}
          cx={radius + padding}
          cy={radius + padding}
        />
      </svg>
      <p
        className={`${styles.value} react-progress-spinner-value`}
        aria-hidden="true"
      >
        <span data-testid="spinner-value">{percentage}</span>
        <span
          className={`${styles.percentage} react-progress-spinner-percentage`}
        >
          %
        </span>
      </p>
    </div>
  );
};

export default Spinner;
