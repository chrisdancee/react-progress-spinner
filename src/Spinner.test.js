import renderer from "react-test-renderer";
import { getNodeText } from "@testing-library/dom";
import { render, screen } from "@testing-library/react";

import Spinner from "./Spinner";

// Snapshots to avoid portentially flakey and cumbersome tests around attribute calculations (strokeDashArray, strokeDashoffset etc.)
test("renders correctly", async () => {
  expect(
    renderer
      .create(<Spinner percentage={33} radius={200} thickness={12} />)
      .toJSON()
  ).toMatchSnapshot();

  expect(
    renderer
      .create(<Spinner percentage={50} radius={150} thickness={8} />)
      .toJSON()
  ).toMatchSnapshot();

  expect(
    renderer
      .create(<Spinner percentage={75} radius={130} thickness={6} />)
      .toJSON()
  ).toMatchSnapshot();
});

test("shows the correct percentage value", async () => {
  render(<Spinner percentage={50} />);
  const spinner = await screen.getByTestId("spinner-value");
  expect(getNodeText(spinner)).toBe("50");
});

test("shows the correct accessibility values", async () => {
  const { rerender } = render(<Spinner percentage={50} />);
  const spinner = await screen.getByTestId("spinner");

  expect(spinner).toHaveAttribute("aria-busy", "true");
  expect(spinner).toHaveAttribute("aria-valuenow", "50");

  rerender(<Spinner percentage={100} />);
  expect(spinner).toHaveAttribute("aria-busy", "false");
  expect(spinner).toHaveAttribute("aria-valuenow", "100");
});
