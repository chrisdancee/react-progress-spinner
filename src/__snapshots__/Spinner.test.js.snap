// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-busy={true}
  aria-live="assertive"
  aria-valuenow={33}
  className="undefined react-progress-spinner-container"
  data-testid="spinner"
  role="progressbar"
  style={
    Object {
      "fontSize": "200px",
    }
  }
>
  <svg
    aria-hidden="true"
    className="undefined react-progress-spinner-spinner"
    height={412}
    viewBox="0 0 412 412"
    width={412}
  >
    <circle
      className="undefined react-progress-spinner-spinner"
      cx={206}
      cy={206}
      r={206}
    />
    <circle
      className="undefined react-progress-spinner-progress"
      cx={206}
      cy={206}
      r={200}
      style={
        Object {
          "strokeDasharray": 1256.6370614359173,
          "strokeDashoffset": 841.9468311620647,
          "strokeWidth": 12,
        }
      }
    />
    <circle
      className="undefined react-progress-spinner-inset"
      cx={206}
      cy={206}
      r={194}
    />
  </svg>
  <p
    aria-hidden="true"
    className="undefined react-progress-spinner-value"
  >
    <span
      data-testid="spinner-value"
    >
      33
    </span>
    <span
      className="undefined react-progress-spinner-percentage"
    >
      %
    </span>
  </p>
</div>
`;

exports[`renders correctly 2`] = `
<div
  aria-busy={true}
  aria-live="assertive"
  aria-valuenow={50}
  className="undefined react-progress-spinner-container"
  data-testid="spinner"
  role="progressbar"
  style={
    Object {
      "fontSize": "150px",
    }
  }
>
  <svg
    aria-hidden="true"
    className="undefined react-progress-spinner-spinner"
    height={308}
    viewBox="0 0 308 308"
    width={308}
  >
    <circle
      className="undefined react-progress-spinner-spinner"
      cx={154}
      cy={154}
      r={154}
    />
    <circle
      className="undefined react-progress-spinner-progress"
      cx={154}
      cy={154}
      r={150}
      style={
        Object {
          "strokeDasharray": 942.4777960769379,
          "strokeDashoffset": 471.23889803846896,
          "strokeWidth": 8,
        }
      }
    />
    <circle
      className="undefined react-progress-spinner-inset"
      cx={154}
      cy={154}
      r={146}
    />
  </svg>
  <p
    aria-hidden="true"
    className="undefined react-progress-spinner-value"
  >
    <span
      data-testid="spinner-value"
    >
      50
    </span>
    <span
      className="undefined react-progress-spinner-percentage"
    >
      %
    </span>
  </p>
</div>
`;

exports[`renders correctly 3`] = `
<div
  aria-busy={true}
  aria-live="assertive"
  aria-valuenow={75}
  className="undefined react-progress-spinner-container"
  data-testid="spinner"
  role="progressbar"
  style={
    Object {
      "fontSize": "130px",
    }
  }
>
  <svg
    aria-hidden="true"
    className="undefined react-progress-spinner-spinner"
    height={266}
    viewBox="0 0 266 266"
    width={266}
  >
    <circle
      className="undefined react-progress-spinner-spinner"
      cx={133}
      cy={133}
      r={133}
    />
    <circle
      className="undefined react-progress-spinner-progress"
      cx={133}
      cy={133}
      r={130}
      style={
        Object {
          "strokeDasharray": 816.8140899333462,
          "strokeDashoffset": 204.20352248333654,
          "strokeWidth": 6,
        }
      }
    />
    <circle
      className="undefined react-progress-spinner-inset"
      cx={133}
      cy={133}
      r={127}
    />
  </svg>
  <p
    aria-hidden="true"
    className="undefined react-progress-spinner-value"
  >
    <span
      data-testid="spinner-value"
    >
      75
    </span>
    <span
      className="undefined react-progress-spinner-percentage"
    >
      %
    </span>
  </p>
</div>
`;
